primitives/src/unique_vec.rs

```diff
@@ -0,0 +1,363 @@
+//! Module with [`UniqueVec`] type and related functional.
+
+#[cfg(not(feature = "std"))]
+use alloc::{borrow::ToOwned as _, format, string::String, vec::Vec};
+use core::{
+    borrow::Borrow,
+    fmt::{Debug, Display},
+    ops::Deref,
+};
+
+use iroha_schema::IntoSchema;
+use parity_scale_codec::{Decode, Encode};
+use serde::{Deserialize, Serialize};
+
+/// Creates a [`UniqueVec`](unique_vec::UniqueVec) from a list of values.
+///
+/// Works like [`vec!`] macro, but does not accept syntax for repeated values
+/// and might return [`Result`].
+#[macro_export]
+macro_rules! unique_vec {
+    () => {
+        $crate::unique_vec::UniqueVec::new()
+    };
+    ($($x:expr),+ $(,)?) => {{
+        let mut v = $crate::unique_vec::UniqueVec::new();
+        $(v.push($x);)+
+        v
+    }};
+}
+
+/// Wrapper type for [`Vec`] which ensures that all elements are unique.
+#[derive(
+    Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize, Encode, Decode, IntoSchema,
+)]
+pub struct UniqueVec<T: PartialEq>(Vec<T>);
```

Comment:
```suggestion
pub struct UniqueVec<T>(Vec<T>);
```

is this bound required on the type? if you remove it you'll be able to derive more things, like `Default`

Reply:
Imo type should have as little bounds as possible and bounds should be moved to places where actually required. 

For example the same pattern is used across std ([`BTreeMap`](https://doc.rust-lang.org/stable/std/collections/struct.BTreeMap.html)).

Reply:
> Imo type should have as little bounds as possible and bounds should be moved to places where actually required.

Indeed I think the same in most of the cases, but IMO sometimes it makes sence to restrict generic in the type. However I don't this this is the case, will fix